File: src/libqlockfile.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const qlockfile_enums = enums;
const std = @import("std");

/// https://doc.qt.io/qt-6/qlockfile.html
pub const qlockfile = struct {
    /// New constructs a new QLockFile object.
    ///
    /// ``` fileName: []const u8 ```
    pub fn New(fileName: []const u8) QtC.QLockFile {
        const fileName_str = qtc.struct_libqt_string{
            .len = fileName.len,
            .data = fileName.ptr,
        };

        return qtc.QLockFile_new(fileName_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qlockfile.html#fileName)
    ///
    /// ``` self: QtC.QLockFile, allocator: std.mem.Allocator ```
    pub fn FileName(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QLockFile_FileName(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qlockfile.FileName: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qlockfile.html#lock)
    ///
    /// ``` self: QtC.QLockFile ```
    pub fn Lock(self: ?*anyopaque) bool {
        return qtc.QLockFile_Lock(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qlockfile.html#tryLock)
    ///
    /// ``` self: QtC.QLockFile, timeout: i32 ```
    pub fn TryLock(self: ?*anyopaque, timeout: i32) bool {
        return qtc.QLockFile_TryLock(@ptrCast(self), @intCast(timeout));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qlockfile.html#unlock)
    ///
    /// ``` self: QtC.QLockFile ```
    pub fn Unlock(self: ?*anyopaque) void {
        qtc.QLockFile_Unlock(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qlockfile.html#setStaleLockTime)
    ///
    /// ``` self: QtC.QLockFile, staleLockTime: i32 ```
    pub fn SetStaleLockTime(self: ?*anyopaque, staleLockTime: i32) void {
        qtc.QLockFile_SetStaleLockTime(@ptrCast(self), @intCast(staleLockTime));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qlockfile.html#staleLockTime)
    ///
    /// ``` self: QtC.QLockFile ```
    pub fn StaleLockTime(self: ?*anyopaque) i32 {
        return qtc.QLockFile_StaleLockTime(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qlockfile.html#tryLock)
    ///
    /// ``` self: QtC.QLockFile ```
    pub fn TryLock2(self: ?*anyopaque) bool {
        return qtc.QLockFile_TryLock2(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qlockfile.html#isLocked)
    ///
    /// ``` self: QtC.QLockFile ```
    pub fn IsLocked(self: ?*anyopaque) bool {
        return qtc.QLockFile_IsLocked(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qlockfile.html#removeStaleLockFile)
    ///
    /// ``` self: QtC.QLockFile ```
    pub fn RemoveStaleLockFile(self: ?*anyopaque) bool {
        return qtc.QLockFile_RemoveStaleLockFile(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qlockfile.html#error)
    ///
    /// ``` self: QtC.QLockFile ```
    pub fn Error(self: ?*anyopaque) i64 {
        return qtc.QLockFile_Error(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qlockfile.html#dtor.QLockFile)
    ///
    /// Delete this object from C++ memory.
    ///
    /// ``` self: QtC.QLockFile ```
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QLockFile_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qlockfile.html#types
pub const enums = struct {
    pub const LockError = enum {
        pub const NoError: i32 = 0;
        pub const LockFailedError: i32 = 1;
        pub const PermissionError: i32 = 2;
        pub const UnknownError: i32 = 3;
    };
};

File: src/libqlockfile.cpp
#include <QLockFile>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qlockfile.h>
#include "libqlockfile.h"
#include "libqlockfile.hxx"

QLockFile* QLockFile_new(const libqt_string fileName) {
    QString fileName_QString = QString::fromUtf8(fileName.data, fileName.len);
    return new QLockFile(fileName_QString);
}

libqt_string QLockFile_FileName(const QLockFile* self) {
    QString _ret = self->fileName();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

bool QLockFile_Lock(QLockFile* self) {
    return self->lock();
}

bool QLockFile_TryLock(QLockFile* self) {
    return self->tryLock();
}

void QLockFile_Unlock(QLockFile* self) {
    self->unlock();
}

void QLockFile_SetStaleLockTime(QLockFile* self, int staleLockTime) {
    self->setStaleLockTime(static_cast<int>(staleLockTime));
}

int QLockFile_StaleLockTime(const QLockFile* self) {
    return self->staleLockTime();
}

bool QLockFile_IsLocked(const QLockFile* self) {
    return self->isLocked();
}

bool QLockFile_RemoveStaleLockFile(QLockFile* self) {
    return self->removeStaleLockFile();
}

int QLockFile_Error(const QLockFile* self) {
    return static_cast<int>(self->error());
}

bool QLockFile_TryLock1(QLockFile* self, int timeout) {
    return self->tryLock(static_cast<int>(timeout));
}

void QLockFile_Delete(QLockFile* self) {
    delete self;
}
